exercises/custom-set/example.rs

```diff
@@ -42,26 +42,26 @@ impl<T: Ord + Clone> CustomSet<T> {
     }
 
     pub fn intersection(&self, other: &Self) -> CustomSet<T> {
-        CustomSet::new(self.collection
-                           .iter()
-                           .cloned()
-                           .filter(|c| other.contains(c))
-                           .collect())
+        CustomSet::new(&self.collection
+                            .iter()
+                            .cloned()
+                            .filter(|c| other.contains(c))
+                            .collect::<Vec<_>>())
```

Comment:
I wonder if it can be done without the collect. First attempt, note that simply remove it completely will result in:

```
error[E0308]: mismatched types
  --> src/lib.rs:45:24
   |
45 |           CustomSet::new(&self.collection
   |  ________________________^
46 | |                             .iter()
47 | |                             .cloned()
48 | |                             .filter(|c| other.contains(c))
   | |__________________________________________________________^ expected slice, found struct `std::iter::Filter`
   |
   = note: expected type `&[T]`
              found type `&std::iter::Filter<std::iter::Cloned<std::slice::Iter<'_, T>>, [closure@src/lib.rs:48:37: 48:58 other:_]>`
```

Reply:
Yes. `Vec::retain` exists where it only keeps the items which meet the predicate. It does operate on `&mut Vec` though so the `Vec` will need to be cloned.

```rust
fn intersection(&self, other: &Self) -> CustomSet<T> {
    let mut collection = self.collection.clone();
    collection.retain(other.contains)
    CustomSet::new(collection)
}
```

[docs](file:///home/shingtaklam1324/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/share/doc/rust/html/std/vec/struct.Vec.html#method.retain)

As well as that, there seems to be no reason for `|c| other.contains(c)` as it is equivalent to `other.contains` as far as I can tell.

Reply:
I think https://stackoverflow.com/questions/37801375/filter-a-slice-into-a-vector may give some hints as how to answer my question

> `Vec::retain` [...] does operate on `&mut Vec`

Then I think it will not serve the goal (not having to create a temporary variable)

I admit I did not state my goal very precisely, therefore this is my fault that it did not serve the goal.

> As well as that, there seems to be no reason for |c| other.contains(c) as it is equivalent to other.contains as far as I can tell.

True story, but I'm not allowed to make that change in this PR since it's out of scope. Good change to put in another PR.

Reply:
I think what I want is https://stackoverflow.com/questions/34969902/how-to-write-a-rust-function-that-takes-an-iterator

```rust
fn do_things(foos: &[usize]) {
    println!("Thanks for giving me {:?}", foos);
}

fn do_more_things<'a, I>(foos: I) where I: IntoIterator<Item = &'a usize> {
    for foo in foos {
        println!("Thanks for giving me more {}", foo);
    }
}

fn main() {
    let foos = [5, 10];
    
    // Works. Great!
    do_things(&foos);
    // Nice, both of these work!
    do_more_things(&foos);
    do_more_things(foos.iter());
    
    let even_foos: Vec<_> = foos.iter().filter(|f| *f % 2 == 0).cloned().collect();
    // Works too. All right.
    do_things(&even_foos);
    do_more_things(&even_foos);

    // Shall we do it without collect() ?
    //let even_foos2 = foos.iter().filter(|f| *f % 2 == 0).cloned();
    // I don't think there's a way to make this one work.
    //do_things(&even_foos2);
    // This is fine though.
    do_more_things(foos.iter().filter(|f| *f % 2 == 0));
}
```